test-suite/Math/NumberTheory/GaussianIntegersTests.hs

```diff
@@ -15,80 +15,143 @@ module Math.NumberTheory.GaussianIntegersTests
+        k  = integerSquareRoot (unPrime p)
+        bs = [1 .. k]
+        asbs = map (\b' -> ((b' * c) `mod` (unPrime p), b')) bs
+        (a, b) = head [ (a', b') | (a', b') <- asbs, a' <= k]
+    in a :+ b
+
+findPrimeProperty1 :: PrimeWrapper Integer -> Bool
+findPrimeProperty1 p'@(PrimeWrapper p)
+  = unPrime p `mod` 4 /= (1 :: Integer)
+  || p1 == p2
+  || abs (p1 * p2) == fromInteger (unPrime p)
+  where
+    p1 = findPrimeReference p'
+    p2 = findPrime' (unPrime p)
 
 -- | Number is prime iff it is non-zero
 --   and has exactly one (non-unit) factor.
-isPrimeProperty :: Integer -> Integer -> Bool
-isPrimeProperty x y
-  =  x == 0 && y == 0
+isPrimeProperty :: GaussianInteger -> Bool
+isPrimeProperty g
+  =  g == 0
   || isPrime g && n == 1
   || not (isPrime g) && n /= 1
   where
-    g = x :+ y
     factors = factorise g
-    nonUnitFactors = filter (\(p, _) -> norm p /= 1) factors
     -- Count factors taking into account multiplicity
-    n = sum $ map snd nonUnitFactors
+    n = sum $ map snd factors
+
+primesSpecialCase1 :: Assertion
+primesSpecialCase1 = assertEqual "primes"
+  (f [1+ι,3,1+2*ι,2+ι,7,11,3+2*ι,2+3*ι,4+ι,1+4*ι,19,23,2+5*ι,5+2*ι,31,6+ι,1+6*ι,5+4*ι,4+5*ι,43,47,2+7*ι,7+2*ι,59,5+6*ι,6+5*ι,67,71,3+8*ι,8+3*ι,79,83,5+8*ι,8+5*ι,4+9*ι,9+4*ι,1+10*ι,10+ι,103,107,3+10*ι,10+3*ι,8+7*ι,7+8*ι,127,131,11+4*ι,4+11*ι,139,10+7*ι,7+10*ι])
+  (f $ take 51 primes)
+  where
+    f :: [GaussianInteger] -> [[GaussianInteger]]
+    f = map sort . groupBy (\g1 g2 -> norm g1 == norm g2)
 
 -- | The list of primes should include only primes.
 primesGeneratesPrimesProperty :: NonNegative Int -> Bool
 primesGeneratesPrimesProperty (NonNegative i) = isPrime (primes !! i)
 
 -- | signum and abs should satisfy: z == signum z * abs z
-signumAbsProperty :: Integer -> Integer -> Bool
-signumAbsProperty x y = z == signum z * abs z
-  where
-    z = x :+ y
+signumAbsProperty :: GaussianInteger -> Bool
+signumAbsProperty z = z == signum z * abs z
 
 -- | abs maps a Gaussian integer to its associate in first quadrant.
-absProperty :: Integer -> Integer -> Bool
-absProperty x y = isOrigin || (inFirstQuadrant && isAssociate)
+absProperty :: GaussianInteger -> Bool
+absProperty z = isOrigin || (inFirstQuadrant && isAssociate)
   where
-    z = x :+ y
     z'@(x' :+ y') = abs z
     isOrigin = z' == 0 && z == 0
     inFirstQuadrant = x' > 0 && y' >= 0     -- first quadrant includes the positive real axis, but not the origin or the positive imaginary axis
-    isAssociate = z' `elem` map (\e -> z * (0 :+ 1) .^ e) [0 .. 3]
+    isAssociate = z' `elem` map (\e -> z * (0 :+ 1) ^ e) [0 .. 3]
+
+gcdGProperty1 :: GaussianInteger -> GaussianInteger -> Bool
+gcdGProperty1 z1 z2
+  = z == 0
+  || z1 `remG` z == 0 && z2 `remG` z == 0 && z == abs z
+  where
+    z = gcdG z1 z2
+
+gcdGProperty2 :: GaussianInteger -> GaussianInteger -> GaussianInteger -> Bool
+gcdGProperty2 z z1 z2
+  = z == 0
+  || gcdG z1' z2' `remG` z == 0
```

Comment:
Is this a bug? How can the GCD be zero? (Or am I reading this wrong?)

Reply:
I don't think it's asking of the `gcd` is zero, but if the remainder of the division of that `gcd` by `z` is 0.

Reply:
It is convenient to set `gcd 0 0` equal to `0`.

Reply:
I see, I was parsing the code incorrectly.